### ISSUE TITLE
Error: Error applying plan:  1 error(s) occurred:  * vsphere_virtual_machine.vm_3: cannot locate virtual machine or template with UUID "423908f8-54b3-432c-5f2b-ff6b8aee5a32": ServerFaultCode: The session is not authenticated

### ISSUE THREAD

**Issue by vknemanavar** [terraform-providers/terraform-provider-vsphere#664]
Hi there,
we have single template to provision 3 VMs one of the VM takes 40min to complete the provision when it comes to 3rd VM we are getting error "vsphere_virtual_machine.vm_3: cannot locate virtual machine or template with UUID "423908f8-54b3-432c-5f2b-ff6b8aee5a32": ServerFaultCode: The session is not authenticated"
Appreciate your help

Terraform v0.11.8
+ provider.vsphere v1.6.0
### Affected Resource(s)

Please list the resources as a list, for example:
- `vsphere_virtual_machine`

### Terraform Configuration Files
```
provider "vsphere" {
  user           = "${var.vsphere_user}"
  password       = "${var.vsphere_password}"
  vsphere_server = "${var.vsphere_server}"
  persist_session = true
  # if you have a self-signed cert
  allow_unverified_ssl = true
}

data "vsphere_datacenter" "dc" {
  name = "${var.vm_1_datacenter}"
}

data "vsphere_datastore" "datastore" {
  name          = "${var.vm_1_datastore}"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_compute_cluster" "cluster" {
  name          = "${var.vm_1_cluster}"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_network" "network" {
  name          = "${var.vm_1_network_interface}"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

data "vsphere_virtual_machine" "template" {
  name          = "${var.vm_1_template}"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}

resource "vsphere_virtual_machine" "vm_1" {
  name             = "${var.vm_1_vmname}"
  folder           = "${var.vm_1_folder}"
  resource_pool_id = "${data.vsphere_compute_cluster.cluster.resource_pool_id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"

  num_cpus = "${var.vm_1_number_of_vcpu}"
  memory   = "${var.vm_1_memory}"
  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  cpu_hot_add_enabled = true
  memory_hot_add_enabled = true
  enable_logging = true
  sync_time_with_host = true

  scsi_type = "${data.vsphere_virtual_machine.template.scsi_type}"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
     label            = "${var.vm_1_vmname}"
#    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
     size             = "${var.vm_1_disk_size}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

customize {
	linux_options {
        host_name = "${var.vm_1_hostname}"
        domain    = "${var.vm_1_domain}"
      }

 network_interface {
  	ipv4_address = "${var.vm_1_ipv4_address}"
        ipv4_netmask = "${var.vm_1_ipv4_prefix_length}"
      }

      ipv4_gateway = "${var.vm_1_ipv4_gateway}"
      dns_server_list = ["${var.dns_servers}"]
      dns_suffix_list = ["${var.vm_1_domain}"]
    }
  }
  provisioner "file" {
        destination = "/tmp/provision.sh"
	content = <<EOF
#!/bin/bash
#Script to install chef client and invoke the cookbook to install middleware
exec &> provision.log
sudo mkdir /NFS
sudo mount 192.168.194.98:/nfs /NFS
sudo tee -a /etc/fstab << END
192.168.194.98:/nfs /NFS nfs defaults,_netdev 0 0
END
sudo tee -a /etc/hosts << END
${var.vm_2_ipv4_address}  ${var.vm_2_hostname}.${var.vm_1_domain} ${var.vm_2_hostname}
${var.vm_3_ipv4_address}  ${var.vm_3_hostname}.${var.vm_1_domain} ${var.vm_3_hostname}
END
sudo rpm -ivh /NFS/Chef_Files_VCS/Automation/chef-14.4.56-1.el7.x86_64.rpm
#Reregistering the RHN
sudo  subscription-manager clean
sudo subscription-manager register --username ibmsmd --password Frh74Mz23
sudo subscription-manager attach --auto
#calling chef ROle
#sudo  chef-client  -o "role[db2_super_v11122_rhel_phadr_role]"
EOF
}

connection {
    type     = "ssh"
    user     = "tfuser"
    password = "${var.tfuser_password}"
 }

provisioner "remote-exec" {
   inline = [
    "chmod +x /tmp/provision.sh",
    "/tmp/provision.sh",
]
}
provisioner "chef" {
    environment     = "_default"
    run_list        = ["role[infosphere_super_v115_mdtier_role]"]
    node_name       = "${var.vm_1_hostname}"
    server_url      = "https://192.168.194.105/organizations/smdperf"
    recreate_client = true
    user_name       = "chefadmin"
    user_key        = "${file("Z:\\Chef_Files_VCS\\Automation\\chef\\Linux\\chef\\certs\\chefadmin.pem")}"
    version         = "12.14"
    skip_install    = true

    # If you have a self signed cert on your chef server change this to :verify_none
    ssl_verify_mode = ":verify_none"
  }
}

resource "vsphere_virtual_machine" "vm_2" {
  depends_on       = ["vsphere_virtual_machine.vm_1"]
  name             = "${var.vm_2_vmname}"
  folder           = "${var.vm_1_folder}"
  resource_pool_id = "${data.vsphere_compute_cluster.cluster.resource_pool_id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"

  num_cpus = "${var.vm_2_number_of_vcpu}"
  memory   = "${var.vm_2_memory}"
  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  cpu_hot_add_enabled = true
  memory_hot_add_enabled = true
  enable_logging = true
  sync_time_with_host = true

  scsi_type = "${data.vsphere_virtual_machine.template.scsi_type}"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
     label            = "${var.vm_2_vmname}"
#    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
     size             = "${var.vm_2_disk_size}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

customize {
	linux_options {
        host_name = "${var.vm_2_hostname}"
        domain    = "${var.vm_1_domain}"
      }

 network_interface {
		ipv4_address = "${var.vm_2_ipv4_address}"
        ipv4_netmask = "${var.vm_1_ipv4_prefix_length}"
      }

      ipv4_gateway = "${var.vm_1_ipv4_gateway}"
      dns_server_list = ["${var.dns_servers}"]
      dns_suffix_list = ["${var.vm_1_domain}"]
    }
  }
  provisioner "file" {
        destination = "/tmp/provision.sh"
	content = <<EOF
	#!/bin/bash
#Script to install chef client and invoke the cookbook to install middleware
exec &> provision.log
sudo mkdir /NFS
sudo mount 192.168.194.98:/nfs /NFS
sudo tee -a /etc/fstab << END
192.168.194.98:/nfs /NFS nfs defaults,_netdev 0 0
END
sudo tee -a /etc/hosts << END
${var.vm_1_ipv4_address}  ${var.vm_1_hostname}.${var.vm_1_domain} ${var.vm_1_hostname}
${var.vm_3_ipv4_address}  ${var.vm_3_hostname}.${var.vm_1_domain} ${var.vm_3_hostname}
END
sudo rpm -ivh /NFS/Chef_Files_VCS/Automation/chef-14.4.56-1.el7.x86_64.rpm
#Reregistering the RHN
sudo  subscription-manager clean
sudo subscription-manager register --username ibmsmd --password Frh74Mz23
sudo subscription-manager attach --auto
#calling chef ROle
#sudo  chef-client  -o "role[db2_super_v11122_rhel_shadr_role]"
EOF
}

connection {
    type     = "ssh"
    user     = "tfuser"
    password = "${var.tfuser_password}"
 }

provisioner "remote-exec" {
   inline = [
    "touch /tmp/metadata_node.out",
    "echo \"${var.vm_1_hostname}.${var.vm_1_domain}\" > /tmp/metadata_node.out",
    "chmod +x /tmp/provision.sh",
    "/tmp/provision.sh",
]
}
provisioner "chef" {
    environment     = "_default"
    run_list        = ["role[infosphere_super_v115_service_tier_role]"]
    node_name       = "${var.vm_2_hostname}"
    server_url      = "https://192.168.194.105/organizations/smdperf"
    recreate_client = true
    user_name       = "chefadmin"
    user_key        = "${file("Z:\\Chef_Files_VCS\\Automation\\chef\\Linux\\chef\\certs\\chefadmin.pem")}"
    version         = "12.14"
    skip_install    = true

    # If you have a self signed cert on your chef server change this to :verify_none
    ssl_verify_mode = ":verify_none"
  }

}

resource "vsphere_virtual_machine" "vm_3" {
  depends_on       = ["vsphere_virtual_machine.vm_2"]
  name             = "${var.vm_3_vmname}"
  folder           = "${var.vm_1_folder}"
  resource_pool_id = "${data.vsphere_compute_cluster.cluster.resource_pool_id}"
  datastore_id     = "${data.vsphere_datastore.datastore.id}"

  num_cpus = "${var.vm_3_number_of_vcpu}"
  memory   = "${var.vm_3_memory}"
  guest_id = "${data.vsphere_virtual_machine.template.guest_id}"

  cpu_hot_add_enabled = true
  memory_hot_add_enabled = true
  enable_logging = true
  sync_time_with_host = true

  scsi_type = "${data.vsphere_virtual_machine.template.scsi_type}"

  network_interface {
    network_id   = "${data.vsphere_network.network.id}"
    adapter_type = "${data.vsphere_virtual_machine.template.network_interface_types[0]}"
  }

  disk {
     label            = "${var.vm_3_vmname}"
#    size             = "${data.vsphere_virtual_machine.template.disks.0.size}"
     size             = "${var.vm_3_disk_size}"
    thin_provisioned = "${data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
  }

  clone {
    template_uuid = "${data.vsphere_virtual_machine.template.id}"

customize {
	linux_options {
        host_name = "${var.vm_3_hostname}"
        domain    = "${var.vm_1_domain}"
      }

 network_interface {
		ipv4_address = "${var.vm_3_ipv4_address}"
        ipv4_netmask = "${var.vm_1_ipv4_prefix_length}"
      }

      ipv4_gateway = "${var.vm_1_ipv4_gateway}"
      dns_server_list = ["${var.dns_servers}"]
      dns_suffix_list = ["${var.vm_1_domain}"]
    }
  }
  provisioner "file" {
        destination = "/tmp/provision.sh"
	content = <<EOF
	#!/bin/bash
#Script to install chef client and invoke the cookbook to install middleware
exec &> provision.log
sudo mkdir /NFS
sudo mount 192.168.194.98:/nfs /NFS
sudo tee -a /etc/fstab << END
192.168.194.98:/nfs /NFS nfs defaults,_netdev 0 0
END
sudo tee -a /etc/hosts << END
${var.vm_1_ipv4_address}  ${var.vm_1_hostname}.${var.vm_1_domain} ${var.vm_1_hostname}
${var.vm_2_ipv4_address}  ${var.vm_2_hostname}.${var.vm_1_domain} ${var.vm_2_hostname}
END
sudo rpm -ivh /NFS/Chef_Files_VCS/Automation/chef-14.4.56-1.el7.x86_64.rpm
sudo rpm -e --nodeps libgcc-4.8.5-28.el7_5.1.x86_64 libgcc-4.8.5-28.el7_5.1.i686
sudo rpm -ivh /NFS/Chef_Files_VCS/Automation/libgcc-4.8.2-16.el7*
#Reregistering the RHN
sudo  subscription-manager clean
sudo subscription-manager register --username ibmsmd --password Frh74Mz23
sudo subscription-manager attach --auto
#calling chef ROle
#sudo  chef-client  -o "role[db2_super_v11122_rhel_shadr_role]"
EOF
}

connection {
    type     = "ssh"
    user     = "tfuser"
    password = "${var.tfuser_password}"
 }

provisioner "remote-exec" {
   inline = [
    "touch /tmp/service_node.out",
    "echo \"${var.vm_2_hostname}.${var.vm_1_domain}\" > /tmp/service_node.out",
    "chmod +x /tmp/provision.sh",
    "/tmp/provision.sh",
]
}
provisioner "chef" {
    environment     = "_default"
    run_list        = ["role[infosphere_super_v115_engine_tier_role]"]
    node_name       = "${var.vm_3_hostname}"
    server_url      = "https://192.168.194.105/organizations/smdperf"
    recreate_client = true
    user_name       = "chefadmin"
    user_key        = "${file("Z:\\Chef_Files_VCS\\Automation\\chef\\Linux\\chef\\certs\\chefadmin.pem")}"
    version         = "12.14"
    skip_install    = true

    # If you have a self signed cert on your chef server change this to :verify_none
    ssl_verify_mode = ":verify_none"
  }
}
```

### Panic Output
Error: Error applying plan:

1 error(s) occurred:

* vsphere_virtual_machine.vm_3: cannot locate virtual machine or template with UUID "423908f8-54b3-432c-5f2b-ff6b8aee5a32": ServerFaultCode: The session is not authenticated

### Expected Behavior

After the second VM is successful it should have pvovisioned 3rd VM

### Actual Behavior

3rd VM unable to provision getting server fault error

### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

**Comment by vknemanavar** [comment]
Does any one has any clue how to resolve this issue
we have single template to provision 3 VMs one of the VM takes 40min to complete the provision when it comes to 3rd VM we are getting error "vsphere_virtual_machine.vm_3: cannot locate virtual machine or template with UUID "423908f8-54b3-432c-5f2b-ff6b8aee5a32": ServerFaultCode: The session is not authenticated"
Appreciate your help

**Comment by vknemanavar** [comment]
@bill-rich Can I know till enhancement to be made what would be the workaround to make this work?

**Comment by bill-rich** [comment]
Hi @vknemanavar! The reason you're running into this issue is that the API sessions are timed out after 30 minutes of inactivity. In your case, it looks like customization on `vsphere_virtual_machine.vm_3` is taking longer than 30 minutes, so the session is timed out. In other long running cases, it is less of a problem because there is some API activity to reset the timer. 

I'm looking at different possibilities for how this can be resolved. I don't have an date on when that enhancement will be ready as it depends some on which implementation we go with. I do have it on the backlog and will be making a decision on how best to handle it soon. I will make sure to keep this issue updated.

If you need to get around this in the meantime, you could use `null` provider to act as a timer and a resource like `vsphere_file` keep the session active. It creates an unnecessary file, and is definitely a hack, but it should give you a workaround until a proper fix is ready. Something like this should work:
```
resrouce "null_resource" "null" {
  depends_on = ["vsphere_virtual_machine.vm_2"]
  provisioner "local-exec" {
    command = "sleep 1200"
  }
}

resource "vsphere_file" "vm_file1" {
  depends_on = ["null_resource.null"]
  ...
}

**Comment by bill-rich** [comment]
This should be resolved by #792 and was released in v1.12.0
